test/gtest/ucs/test_datatype.cc

```diff
@@ -1356,3 +1359,358 @@ UCS_MT_TEST_F(test_datatype_ptr_map_safe, safe_put, num_threads)
+    ucs_linear_func_t segment_func = ucs_linear_func_make(1, 1);
+    check_signle_segment_func(func, segment_func);
+
+    for (size_t point = SIZE_MAX; point != 0; point /= 2) {
+        compare_point_value(zero_func, UCS_LINEAR_FUNC_ZERO, point);
+        compare_point_value(func, segment_func, point);
+    }
+
+    ucs_piecewise_func_free(&zero_func);
+    ucs_piecewise_func_free(&func);
+}
+
+UCS_TEST_F(test_piecewise_func, segment_add) {
+    std::vector<std::pair<size_t, size_t>> ranges_to_test =
+            {{0, 0},   {0, 5},         {0, SIZE_MAX},       {10, 20},
+             {20, 20}, {20, SIZE_MAX}, {SIZE_MAX, SIZE_MAX}};
+    ucs_linear_func_t initial_segment_func = ucs_linear_func_make(1, 1);
+    ucs_linear_func_t new_segment_func     = ucs_linear_func_make(2, 2);
+    ucs_linear_func_t result_func = ucs_linear_func_add(new_segment_func,
+                                                        initial_segment_func);
+
+    for (const auto &range : ranges_to_test) {
+        segment seg = {range.first, range.second, new_segment_func};
+        UCS_TEST_MESSAGE << get_segment_string(seg);
+
+        ucs_piecewise_func_t func;
+        ucs_piecewise_func_init(&func, 1, 1);
+        piecewise_func_add_segment(&func, seg.start, seg.end, seg.func);
+
+        /* Check number of the segments in the result function */
+        size_t expected_num_segments = 3;
+        if (seg.start == 0) {
+            --expected_num_segments;
+        }
+        if (seg.end == SIZE_MAX) {
+            --expected_num_segments;
+        }
+        ASSERT_EQ(get_num_segments(func), expected_num_segments);
+
+        /* Check points which should be affected by added segment */
+        for (auto point : get_segment_points(seg)) {
+            compare_point_value(func, result_func, point);
+        }
+
+        /* Check points which should not be affected by added segment */
+        std::vector<size_t> initial_segment_points;
+        if (seg.start > 0) {
+            initial_segment_points.emplace_back(seg.start - 1);
+        }
+        if (seg.end < SIZE_MAX) {
+            initial_segment_points.emplace_back(seg.end + 1);
+        }
+
+        for (auto point : initial_segment_points) {
+            compare_point_value(func, initial_segment_func, point);
+        }
+
+        ucs_piecewise_func_free(&func);
+    }
+}
+
+UCS_TEST_F(test_piecewise_func, random_segment_add) {
+    constexpr size_t segmetns_num = 10;
+    constexpr size_t test_num     = 100;
+
+    for (size_t test_it = 0; test_it < test_num; ++test_it) {
+        ucs_piecewise_func_t func;
+        ucs_piecewise_func_init(&func, 0, 0);
+
+        std::map<size_t, double> points_to_values;
+        for (auto point : {(size_t)0, SIZE_MAX / 2, SIZE_MAX}) {
+            points_to_values.insert(
+                    {point, ucs_piecewise_func_apply(&func, point)});
+        }
+
+        std::vector<segment> segments;
+        for (size_t i = 0; i < segmetns_num; ++i) {
```

Comment:
typo

Reply:
This PR contains some outdated code state. This is done in this PR (https://github.com/openucx/ucx/pull/9569) which should be merged first.